### DISCUSSION TITLE
what is the maximum support version of node.js and alpine can build code-server with docker image

### DISCUSSION THREAD

**Discussion by stephenwongcandr** [coder/code-server#6034] [Q&A]
I was using [node](node:14.4.0-alpine3.11) as base image with code server v3.4.1, it was work fine,
but since the node 14 is end of lifecycle , and there have some security concern,
I am trying to build a new image with code server,
but currently I was building fail with node:18.14.0-alpine3.17 / node:18.14.0-alpine3.16 / node:16.15.1-alpine3.16 with v.4.9.0 or v.4.10.0 code server,
can anyone recommend any node.js 16/18 version and alpine 3.16+ version can support any version's code server?

**Comment by code-asher** [comment]
~I think this may be difficult since the versions are not always backwards compatible.~

~Do you have to use 3.4.1?  The latest code-server (4.10.0) uses Node v16 (we use both 16.13.0 and latest v16).~

nvm, see below

**Reply by code-asher** [reply]
Oh sorry I missed that you were already building with 4.9.0 and 4.10.0.

Could you post the error logs?  We might be able to figure it out that way.

**Reply by stephenwongcandr** [reply]
hi, sorry for late reply,
here is the error log after run the docker container,
And I am using base image node:16.15.1-alpine3.16 to install the code server from https://github.com/cdr/code-server/releases/download/v4.10.0/code-server-4.10.0-linux-arm64.tar.gz

`node:internal/modules/cjs/loader:1189
  return process.dlopen(module, path.toNamespacedPath(filename));
                 ^

Error: Error loading shared library ld-linux-aarch64.so.1: No such file or directory (needed by /.local/lib/code-server-4.10.0-linux-arm64/node_modules/argon2/lib/binding/napi-v3/argon2.node)
    at Object.Module._extensions..node (node:internal/modules/cjs/loader:1189:18)
    at Module.load (node:internal/modules/cjs/loader:981:32)
    at Function.Module._load (node:internal/modules/cjs/loader:822:12)
    at Module.require (node:internal/modules/cjs/loader:1005:19)
    at require (node:internal/modules/cjs/helpers:102:18)
    at Object.<anonymous> (/.local/lib/code-server-4.10.0-linux-arm64/node_modules/argon2/argon2.js:9:25)
    at Module._compile (node:internal/modules/cjs/loader:1105:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1159:10)
    at Module.load (node:internal/modules/cjs/loader:981:32)
    at Function.Module._load (node:internal/modules/cjs/loader:822:12) {
  code: 'ERR_DLOPEN_FAILED'
}`

**Reply by code-asher** [reply]
Ahhh the prebuilt `linux-` builds are not compatible with Alpine.  I think you will have to either make your own build or install with npm.

**Reply by stephenwongcandr** [reply]
@code-asher that weird, so I can build the code-server-3.4.1-linux-amd64.tar.gz with node:14.4.0-alpine3.11,
what do you mean about installing with npm?
is that mean I type ```npm install``` during the image build?

**Reply by code-asher** [reply]
Yeah Alpine uses musl not glibc so the precompiled builds on GitHub are not compatible since the native Node modules are compiled against glibc.

I see three options:

1. Build from source (https://github.com/coder/code-server/blob/main/docs/CONTRIBUTING.md#build)
2. Download one of the prebuilds from GitHub (does not matter which, but `linux-amd64` for example), and rebuild all the Node modules.  It should work to do `npm rebuild` in both the root and `lib/vscode` but if that does not work you can delete `node_modules` and run `npm install --legacy-peer-deps --omit=dev` in both directories.
3. Install from npm with `npm install -g code-server --unsafe-perm` (or use our install script which will fall back to npm).  This is similar to option 2, just pulling the compiled source NPM instead of GitHub.

No matter which way you go you will need to have the required dependencies for compilation.

I tried out the npm installation instructions at https://github.com/coder/code-server/blob/main/docs/npm.md#alpine but I ran into issues:

-  I had to install `python3`
- I added `libsecret` so secret storage works
- `lib/vscode/node_modules` was completely missing so I had to manually add those.  An npm bug maybe?  You might not have the same issue.

```console
docker run -it -p 8080 node:16.19.1-alpine3.17 sh
apk add alpine-sdk bash libstdc++ libc6-compat python3 libsecret
npm install -g code-server --unsafe-perm
cd /usr/local/lib/node_modules/code-server/lib/vscode
npm install --legacy-peer-deps
code-server --auth none --host 0.0.0.0
```

We have an open issue for adding Alpine releases but I am not sure we can get to it any time soon: https://github.com/coder/code-server/issues/3431
